### PR TITLE
Rework Source structure

### PR DESCRIPTION
**Pull request by melissa-barca**
This PR creates the basic structure to allow for a user to dynamically create and remove Source Panes within the main window. It is the first step of this process and adds no functional changes. 

`SourceShim` and `SourceBuildHelper` were classes previously used in initializing `Source` and `SourcePane` that required initialization information to be known at compile time which is no longer feasible as users will be able to add/remove columns. These classes have been removed with their logic being moved to `Source.java` and `SourcePane.java`. 

This PR has been through a preliminary QA review. It is related to the set of enhancements mentioned in #6428.

### CODE REVIEW

File: src/gwt/src/org/rstudio/studio/client/workbench/views/environment/EnvironmentPresenter.java [hunk 1]
```diff
@@ -83,7 +83,6 @@
 import org.rstudio.studio.client.workbench.views.environment.model.RObject;
```
Comment (gtritchie) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/environment/EnvironmentPresenter.java:83]: update file copyright year

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/Source.java [hunk 2]
```diff
@@ -14,6 +14,7 @@
  */
 package org.rstudio.studio.client.workbench.views.source;
 
+import com.google.gwt.core.client.GWT;
```
Comment (gtritchie) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/Source.java:17]: not used

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/Source.java [hunk 3]
```diff
@@ -409,40 +413,48 @@ public Source(Commands commands,
...
          }
       });
       
-      events.addHandler(SourceOnSaveChangedEvent.TYPE, 
+      events_.addHandler(NewDocumentWithCodeEvent.TYPE,
+                         new NewDocumentWithCodeEvent.Handler() {
+         @Override
+         public void onNewDocumentWithCode(NewDocumentWithCodeEvent event)
+         {
+            onNewDocumentWithCode(event);
```
Comment (gtritchie) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/Source.java:453]: Doesn't this trigger an infinite recursive loop? (onNewDocumentWithCode calling itself?).

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/Source.java [hunk 4]
```diff
@@ -464,15 +476,15 @@ public void onSwitchToDoc(SwitchToDocEvent event)
...
+      events_.addHandler(SourceFileSavedEvent.TYPE, new SourceFileSavedHandler()
       {
          public void onSourceFileSaved(SourceFileSavedEvent event)
          {
             pMruList_.get().add(event.getPath());
          }
       });
-      
-      events.addHandler(SourcePathChangedEvent.TYPE, 
+
```
Comment (gtritchie) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/Source.java:486]: extra whitespace

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/Source.java [hunk 5]
```diff
@@ -518,8 +530,8 @@ public void onSourceNavigation(SourceNavigationEvent event)
          }
       });
       
-      events.addHandler(SourceExtendedTypeDetectedEvent.TYPE, this);
-      
+      events_.addHandler(SourceExtendedTypeDetectedEvent.TYPE, this);
+
```
Comment (gtritchie) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/Source.java:534]: extra whitespace

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/Source.java [hunk 6]
```diff
@@ -557,8 +569,8 @@ public void execute(EditingTarget editor)
                });
          }
       });
-      
-      events.addHandler(CollabEditEndedEvent.TYPE, 
+
```
Comment (gtritchie) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/Source.java:572]: whitespace

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/Source.java [hunk 7]
```diff
@@ -575,8 +587,8 @@ public void execute(EditingTarget editor)
                });
          }
       });
-      
-      events.addHandler(NewWorkingCopyEvent.TYPE, 
+
```
Comment (gtritchie) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/Source.java:590]: whitespace

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/Source.java [hunk 8]
```diff
@@ -688,14 +727,61 @@ else if (userPrefs_.editorKeybindings().getValue() == UserPrefs.EDITOR_KEYBINDIN
...
       if (SourceWindowManager.isMainSourceWindow())
       {
-         openProjectDocs(session);
+         openProjectDocs(session_);
          openEditPublishedDocs();
       }
       
       // add vim commands
       initVimCommands();
+      ensureVisible(false);
```
Comment (gtritchie) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/Source.java:736]: Why is this added here?

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/Source.java [hunk 9]
```diff
@@ -688,14 +727,61 @@ else if (userPrefs_.editorKeybindings().getValue() == UserPrefs.EDITOR_KEYBINDIN
...
    }
    
+   public void withSaveFilesBeforeCommand(final Command command,
```
Comment (gtritchie) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/Source.java:739]: Brief JavaDoc comment about method's purpose would be good.

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/Source.java [hunk 10]
```diff
@@ -688,14 +727,61 @@ else if (userPrefs_.editorKeybindings().getValue() == UserPrefs.EDITOR_KEYBINDIN
...
+                        }
+
+                        handleUnsavedChangesBeforeExit(
+                                              result.getSaveTargets(),
+                                              command);
+
+                     }
+                   },
+                   cancelCommand
+            ).showModal();
```
Comment (gtritchie) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/Source.java:776]: lambda:

```
            new UnsavedChangesDialog(
                  commandSource,
                  alwaysSaveOption,
                  unsavedSourceDocs,
                  dialogResult ->
                  {
                     if (dialogResult.getAlwaysSave())
                     {
                        userPrefs_.saveFilesBeforeBuild().setGlobalValue(true);
                        userPrefs_.writeUserPrefs();
                     }
   
                     handleUnsavedChangesBeforeExit(
                                           dialogResult.getSaveTargets(),
                                           command);
   
                  },
                  cancelCommand
            ).showModal();
```

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/Source.java [hunk 11]
```diff
@@ -2577,6 +2662,21 @@ public void saveWithPrompt(UnsavedChangesTarget target,
...
+         @Override
+         public void execute()
+         {
+            handleUnsavedChangesBeforeExit(
+                                 new ArrayList<UnsavedChangesTarget>(),
+                                 onCompleted);
+         }
+
+      };
+   }
```
Comment (gtritchie) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/Source.java:2679]: Reduce this whole method to:

```
 public Command revertUnsavedChangesBeforeExitCommand(final Command onCompleted)
   {
      return () -> handleUnsavedChangesBeforeExit(new ArrayList<>(), onCompleted);
   }
```

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/SourceSatellitePresenter.java [hunk 12]
```diff
@@ -29,7 +29,7 @@
    {}
```
Comment (gtritchie) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/SourceSatellitePresenter.java:29]: update file copyright year

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/SourceSatelliteWindow.java [hunk 13]
```diff
@@ -49,6 +49,7 @@ public SourceSatelliteWindow(Provider<EventBus> pEventBus,
                                 Provider<SourceWindow> pSourceWindow,
```
Comment (gtritchie) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/SourceSatelliteWindow.java:49]: update file copyright year

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/SourceWindow.java [hunk 14]
```diff
@@ -69,12 +69,12 @@ public SourceWindow(
          Provider<DesktopHooks> pDesktopHooks,
```
Comment (gtritchie) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/SourceWindow.java:69]: update file copyright year

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/SourceWindowManager.java [hunk 15]
```diff
@@ -102,7 +102,7 @@ public SourceWindowManager(
          EventBus events,
```
Comment (gtritchie) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/SourceWindowManager.java:102]: update file header year

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/TextEditingTarget.java [hunk 16]
```diff
@@ -5574,24 +5574,36 @@ public void execute()
...
          events_.fireEvent(new LaunchShinyApplicationEvent(getPath(),
                prefs_.shinyBackgroundJobs().getValue() ?
                   ShinyApplication.BACKGROUND_APP :
                   ShinyApplication.FOREGROUND_APP, getExtendedFileType()));
-      }, "Run Shiny Application");
+      },
+      new Command() {
+         public void execute()
+         {
+         }
```
Comment (gtritchie) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/TextEditingTarget.java:5587]: labmda: `() -> {}`

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/TextEditingTarget.java [hunk 17]
```diff
@@ -5574,24 +5574,36 @@ public void execute()
...
          public void execute()
          {
             events_.fireEvent(new LaunchPlumberAPIEvent(getPath()));
          }
-      }, "Run Plumber API");
+      },
+      new Command() {
+         public void execute()
+         {
+         }
```
Comment (gtritchie) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/TextEditingTarget.java:5604]: lambda: `() -> {}`

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/Source.java [hunk 18]
```diff
@@ -586,17 +598,57 @@ public void onNewWorkingCopy(NewWorkingCopyEvent event)
...
+      events_.addHandler(OpenProfileEvent.TYPE, this);
+      events_.addHandler(RequestDocumentSaveEvent.TYPE, this);
+      events_.addHandler(RequestDocumentCloseEvent.TYPE, this);
+      
+      initialized_ = true;
+   }
+
+   public void loadFullSource()
+   {
+      AceEditor.preload();
```
Comment (kevinushey) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/Source.java:618]: This might not be safe -- if I understand correctly, this call will load Ace concurrently, but it seems like since we're about to initialize tabs (which might require Ace to be available) we would want to ensure Ace has really been loaded.

File: src/gwt/src/org/rstudio/studio/client/workbench/prefs/model/UserPrefs.java [hunk 19]
```diff
@@ -100,6 +100,7 @@ public void writeUserPrefs()
 
    public void writeUserPrefs(CommandWithArg<Boolean> onCompleted)
    {
+      UpdatePrefs(session_.getSessionInfo().getPrefs());
```
Comment (kevinushey) [inline src/gwt/src/org/rstudio/studio/client/workbench/prefs/model/UserPrefs.java:103]: nit: method name should be camelCase?

Comment (kevinushey) [reply src/gwt/src/org/rstudio/studio/client/workbench/prefs/model/UserPrefs.java:103]: (realize this is a pre-existing method but maybe worth changing while you're working with this code)

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/Source.java [hunk 20]
```diff
@@ -586,17 +598,57 @@ public void onNewWorkingCopy(NewWorkingCopyEvent event)
...
+      if (userPrefs_.editorKeybindings().getValue() == UserPrefs.EDITOR_KEYBINDINGS_VIM)
+         ShortcutManager.INSTANCE.setEditorMode(KeyboardShortcut.MODE_VIM);
+      else if (userPrefs_.editorKeybindings().getValue() == UserPrefs.EDITOR_KEYBINDINGS_EMACS)
+         ShortcutManager.INSTANCE.setEditorMode(KeyboardShortcut.MODE_EMACS);
+      else if (userPrefs_.editorKeybindings().getValue() == UserPrefs.EDITOR_KEYBINDINGS_SUBLIME)
+         ShortcutManager.INSTANCE.setEditorMode(KeyboardShortcut.MODE_SUBLIME);
+      else
+         ShortcutManager.INSTANCE.setEditorMode(KeyboardShortcut.MODE_DEFAULT);
+   
+      events_.fireEvent(new DocTabsChangedEvent(null,
```
Comment (kevinushey) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/Source.java:632]: Can we document the intent here? (why do we need to fire this "empty" DocTabsChangedEvent?)

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/Source.java (hunk 18 above)
Comment (melissa-barca) [reply src/gwt/src/org/rstudio/studio/client/workbench/views/source/Source.java]: Good catch. I updated the code with a new method that loads Ace and calls loadFullSource (now private) in its callback.
